src/OrleansProviders/Streams/Common/SimpleCache/SimpleQueueCache.cs

```diff
@@ -244,40 +242,24 @@ internal bool TryGetNextMessage(SimpleQueueCacheCursor cursor, out IBatchContain
             Log(logger, "TryGetNextMessage: {0}", cursor);
 
             batch = null;
+            if (!cursor.IsSet) return false;
 
-            if (cursor == null) throw new ArgumentNullException("cursor");
-            
-            //if not set, try to set and then get next
-            if (!cursor.IsSet)
-            {
-                InitializeCursor(cursor, cursor.SequenceToken, false);
-                return cursor.IsSet && TryGetNextMessage(cursor, out batch);
-            }
-
-            // has this message been purged
-            if (cursor.SequenceToken.Older(cachedMessages.Last.Value.SequenceToken))
-            {
-                throw new QueueCacheMissException(cursor.SequenceToken, cachedMessages.Last.Value.SequenceToken, cachedMessages.First.Value.SequenceToken);
-            }
-
-            // Cursor now points to a valid message in the cache. Get it!
             // Capture the current element and advance to the next one.
             batch = cursor.Element.Value.Batch;
-            
-            // Advance to next:
+
+            // If we are at the end of the cache unset cursor and move offset one forward
             if (cursor.Element == cachedMessages.First)
             {
-                // If we are at the end of the cache unset cursor and move offset one forward
-                ResetCursor(cursor, ((EventSequenceToken)cursor.SequenceToken).NextSequenceNumber());
+                UnsetCursor(cursor, null);
```

Comment:
Yes.  This is an important part of the change.  We used to always need a sequence token, either of the next message we expect or of some future sequence token we expect to deliver from.  This required that we be able to compute the next sequence token given the current sequence token (the NextSequenceNumber call) which is not functionality supported by sequence tokens.

Now we only expect to have a sequence token if the cursor is currently active or if we're waiting for a future event.  If an active cursor reaches the most recent item in the cache, and has no more items to deliver, it becomes unset and waits for more data, which is signaled by the Refresh call which now takes a new sequence token to start from (rather than starting from the computed one originally left in the cursor).